addons/mail/static/src/messaging/entity/composer/composer.js

```diff
@@ -245,20 +464,66 @@ function ComposerFactory({ Entity }) {
         _reset() {
             this.update({
                 attachments: [['unlink-all']],
+                mentionedPartners: [['unlink-all']],
                 textInputContent: '',
                 textInputCursorStart: 0,
                 textInputCursorEnd: 0,
             });
         }
 
+        /**
+         * Validates user's current typing as a correct mention keyword
+         * in order to trigger mentions suggestions display.
+         * Returns the mention keyword without the delimiter if it has been validated
+         * and false if not
+         *
+         * @private
+         * @param {string} delimiter
+         * @param {boolean} beginningOnly
+         * @returns {string|boolean}
+         */
+        _validateMentionKeyword(delimiter, beginningOnly) {
+            const leftString = this.textInputContent.substring(0, this.textInputCursorStart);
+
+            // use position before delimiter because there should be whitespaces
+            // or line feed/carriage return before the delimiter
+            const beforeDelimiterPosition = leftString.lastIndexOf(delimiter) - 1;
+            if (beginningOnly && beforeDelimiterPosition > 0) {
+                return false;
+            }
+            let searchStr = this.textInputContent.substring(beforeDelimiterPosition, this.textInputCursorStart);
+            // regex string start with delimiter or whitespace then delimiter
+            const pattern = "^"+delimiter+"|^\\s"+delimiter;
+            const regexStart = new RegExp(pattern, 'g');
+            // trim any left whitespaces or the left line feed/ carriage return
+            // at the beginning of the string
+            searchStr = searchStr.replace(/^\s\s*|^[\n\r]/g, '');
+            if (regexStart.test(searchStr) && searchStr.length) {
+                searchStr = searchStr.replace(pattern, '');
+                return !searchStr.includes(' ') && !/[\r\n]/.test(searchStr)
+                    ? searchStr.replace(delimiter, '')
+                    : false;
+            }
+            return false;
+        }
     }
 
     Composer.entityName = 'Composer';
 
     Composer.fields = {
+        activeSuggestedPartner: many2one('Partner'),
         attachments: many2many('Attachment', {
             inverse: 'composers',
         }),
+        isDoFocus: attr({
+            default: false,
+        }),
+        mentionedPartners: many2many('Partner', {
+            compute: '_computeMentionedPartners',
+            dependencies: ['textInputContent'],
+        }),
+        suggestedPartners: many2many('Partner'),
+        suggestedUsers: many2many('Partner'),
```

Comment:
The naming is not correct, it's not a good idea to use "user" naming when a partner is actually referenced.

The name should be more in accordance to its intent ex:
`suggestedPartners` for [0] and `extraSuggestedPartners` for [1]

to brainstorm with @alexkuhn

Reply:
Indeed, this is confusing. Would be easier to get what each does with a better naming, like `mainSuggestedPartners` and `extraSuggestedPartners`? Not sure with these names, to brainstorm vocally maybe? :)

Reply:
It's just a locally scoped variable in the python, its name doesn't really matter.
Note: if we have diff on this method one day, I'll ask "users" to be renamed there too.

But what's more important is that here we are talking about a field on a model, its name has to be well understood outside of context.